Higher-order function application updates to programming paradigms

wilmoore opened the pull request:
- added a section on higher-order function application
- corrected many php manual links (not everyone wants the english version of the manual, etc.)

wilmoore committed:
added a section on higher-order function application

wilmoore committed:
slight modification of wording around higher-order function application for clarity

wilmoore committed:
corrected all of the php manual links in the programming paradigms document as most were direct to a specific mirror instead of allowing the mirror redirection to happen on the fly -- also, not everyone will want the english version of the manual

grakic commented:
Hi @willmoore, can you review existing https://github.com/codeguy/php-the-right-way/pull/67 and compare so we get the best result possible?

wilmoore commented:
Pretty good examples in there; however, I think it suffers from being overly academic in some areas and jumps around a lot. I think there is some overlap between the two. Ultimately, the best guide would be one that starts with simple anonymous functions then goes into higher-order function application as the guide progresses. I purposefully didn't touch the opening piece because I think there should be an intro as well as a more in-depth look at function application.

philsturgeon commented on _posts/03-02-01-Programming-Paradigms.md:
```diff
@@ -30,6 +30,44 @@ $greet('World');
+PHP manual alludes to these being synonymous, this is slightly misleading. A closure is an anonymous function that is capable of capturing the
+current run-time state for later use. Below are examples of anonymous functions and closures used for higher-order function application.
+
+annonymous functions:
```
Capital _"A"_?

wilmoore replied:
@philsturgeon Thanks...fixed.

wilmoore committed:
corrected spelling error and capitalization

wilmoore committed:
Merged issue-67 with my contributed higher-order function section + lots of language/semantic cleanup

wilmoore committed:
utf-8 character checking

philsturgeon commented on _posts/03-02-01-Programming-Paradigms.md:
```diff
@@ -15,7 +15,14 @@ PHP 5.4 (2012).
+PHP supports first-class and higher-order functions. This means that a function literal (or lambda expression or anonymous function)
+can be assigned to a variable or passed as a parameter to another function. Variables containing functions can be invoked on-the-fly.
+
+Anonymous Functions:
```
Should these all be h4's you think? looks a bit odd as just text.

wilmoore replied:
Perhaps we get something out and iterate on the look/feel thereafter. That being said, I'd be happy with you making the heading change as you see fit if  it doesn't look good as-is.

grakic commented:
Hi @wilmoore 

I pushed a new iteration of my earlier pull request taking some of your changes in. It is in issue #67

I also removed `combine` example and instead used `array_filter` for both as you suggested in your comments.

wilmoore commented:
@philsturgeon @grakic This version is a consolidation of #72 and #67. I've organized it so that the information conveys a natural progression through anonymous functions, closures, and higher-order functions. Should be solid enough to merge and we can of course iterator further if necessary, but I think this version is improvement enough over what is posted now to warrant the merge.

wilmoore commented:
BTW, I also cleaned up the links to the manual which were all over the place -- they are now much more consistent. I've also cleaned up the passive-aggressive language around Ruby and method_missing and just conveyed the point directly.

grakic commented:
Your changes on links and meta-programming are non-conflicting, and can be made a new pull request.

But I think that the final version of issue #67 is better over what is suggested here.

Key concepts of functional programming (first-class functions, recursion, higher order functions, followed by anonymous functions and closures) are not stated as clear as in Issue #67. I tried hard to describe these concepts both correctly and to say why would someone want to use them. I also tried to be as short as possible, this is the 3rd level section and should not be given too much space. Interested reader can look up for more info elsewhere after reading the section.

The example is also more clear to follow, with a single topic (just numbers and array_filter, no days, no regexp, no join).

At last, #67 is before #72

wilmoore commented:
@grakic I'm happy to have either to be honest. Whatever works :)

codeguy commented:
If we're potentially dropping down to H4s, I'm not sure this is the best fit for the main page. The content you guys have done is excellent though. Let's keep the very brief/basic description on the home page, and link to a stand-alone tutorial page for this expanded content w/ some code examples.

I haven't had the chance to setup the structure for these stand-alone expanded pages, but I was thinking of creating a `pages/` folder to use as a bucket for stand-alone pages like this. We can link to them from the main document. Each page in the `pages/` folder can use the default layout.

wilmoore commented:
Sounds good to me.

codeguy commented:
Closing. See duplicate #67.
